## YOLO
### Pytorch  기반 물체인식 모델
- CNN , RCNN(Regions with CNN)
- https://github.com/ultralytics/ultralytics
#### YOLOv5 이상 설치
```shell
> pip install ultralytics
```

##### yolo  설치

In [3]:
!pip install ultralytics

   ---------------------------------------- 0.0/974.8 kB ? eta -:--:--
   --------------------------------------- 974.8/974.8 kB 20.5 MB/s eta 0:00:00
   ---------------------------------------- 0.0/39.5 MB ? eta -:--:--
   -- ------------------------------------- 2.4/39.5 MB 76.2 MB/s eta 0:00:01
   ----- ---------------------------------- 5.0/39.5 MB 63.0 MB/s eta 0:00:01
   -------- ------------------------------- 8.1/39.5 MB 65.0 MB/s eta 0:00:01
   ----------- ---------------------------- 11.3/39.5 MB 59.5 MB/s eta 0:00:01
   -------------- ------------------------- 14.5/39.5 MB 65.6 MB/s eta 0:00:01
   ----------------- ---------------------- 17.1/39.5 MB 65.6 MB/s eta 0:00:01
   -------------------- ------------------- 19.9/39.5 MB 59.5 MB/s eta 0:00:01
   ---------------------- ----------------- 22.2/39.5 MB 54.4 MB/s eta 0:00:01
   ------------------------ --------------- 24.0/39.5 MB 50.4 MB/s eta 0:00:01
   -------------------------- ------------- 26.3/39.5 MB 50.1 MB/s eta 


[notice] A new release of pip is available: 24.0 -> 25.0.1
[notice] To update, run: python.exe -m pip install --upgrade pip


##### 콘솔에서 예측

In [4]:
# yolo11n.pt - pretrained YOLO model
!yolo predict model=yolo11n.pt source='https://ultralytics.com/images/bus.jpg'

Creating new Ultralytics Settings v0.0.6 file ✅ 
View Ultralytics Settings with 'yolo settings' or at 'C:\Users\Admin\AppData\Roaming\Ultralytics\settings.json'
Update Settings with 'yolo settings key=value', i.e. 'yolo settings runs_dir=path/to/dir'. For help see https://docs.ultralytics.com/quickstart/#ultralytics-settings.
Ultralytics 8.3.109 🚀 Python-3.11.9 torch-2.6.0+cu118 CUDA:0 (NVIDIA GeForce RTX 4060, 8188MiB)
YOLO11n summary (fused): 100 layers, 2,616,248 parameters, 0 gradients, 6.5 GFLOPs

image 1/1 c:\Source\iot_dataanalysis_2025\day53\bus.jpg: 640x480 4 persons, 1 bus, 71.1ms
Speed: 2.4ms preprocess, 71.1ms inference, 138.7ms postprocess per image at shape (1, 3, 640, 480)
Results saved to C:\Source\iot_dataanalysis_2025\runs\detect\predict
💡 Learn more at https://docs.ultralytics.com/modes/predict



  0%|          | 0.00/5.35M [00:00<?, ?B/s]
100%|██████████| 5.35M/5.35M [00:00<00:00, 65.7MB/s]

  0%|          | 0.00/134k [00:00<?, ?B/s]
100%|██████████| 134k/134k [00:00<00:00, 8.54MB/s]


- 설정 관련
    - C:\Users\Admin\AppData\Roaming\Ultralytics\settings.json
    - 설정을 확인하거나 바꾸려면:
        - 설정 보기: yolo settings
        - 설정 변경: yolo settings 키=값
        - 예: yolo settings runs_dir=c:/your/folder
-  모델 다운로드 :yolo11n.pt라는 모델 파일을 자동으로 다운로드 했어요.
    - 이건 YOLOv8의 초경량 버전 (n = nano).
- 모델 정보
    - YOLO11n 모델은 총 100개의 레이어, 약 260만 개의 파라미터.
    - 처리 속도: 6.5 GFLOPs 정도로 가볍고 빠름.
    - 사용한 GPU: RTX 4060 (8GB)
- 이미지 예측
    - 테스트 이미지 bus.jpg도 자동으로 다운로드했고,
    - 결과: 4명의 사람, 1대의 버스를 감지.
    - 결과 저장 위치: C:\Source\iot_dataanalysis_2025\runs\detect\predict
    - <img src = '../runs\detect\predict/bus.jpg' width =300>


##### 파이썬으로 예측

In [5]:
# YOLO 모듈 로드
from ultralytics import YOLO

In [6]:
# YOLO 클래스가 들어오는 모델의 버전에 따라 알아서 YOLO 예측모델 객체 생성
model = YOLO('./yolo11n.pt')


##### coco8.yaml
- Train/val/test + Classes + 훈련내용
    - https://github.com/ultralytics/ultralytics/blob/main/ultralytics/cfg/datasets/coco8.yaml
- 이미지파일
    - https://github.com/ultralytics/assets/releases/download/v0.0.0/coco8.zip
- https://github.com/ultralytics/ultralytics/blob/main/ultralytics/cfg/datasets/coco8.yaml 내용대로 훈련을 시킨 결과 ->yolo11n.pt

In [7]:
train_results = model.train(
    data ='./coco8.yaml' ,
    epochs = 100,
    imgsz =640,
    device = 'cuda:0'
)

Ultralytics 8.3.109  Python-3.11.9 torch-2.6.0+cu118 CUDA:0 (NVIDIA GeForce RTX 4060, 8188MiB)
engine\trainer: task=detect, mode=train, model=./yolo11n.pt, data=./coco8.yaml, epochs=100, time=None, patience=100, batch=16, imgsz=640, save=True, save_period=-1, cache=False, device=cuda:0, workers=8, project=None, name=train, exist_ok=False, pretrained=True, optimizer=auto, verbose=True, seed=0, deterministic=True, single_cls=False, rect=False, cos_lr=False, close_mosaic=10, resume=False, amp=True, fraction=1.0, profile=False, freeze=None, multi_scale=False, overlap_mask=True, mask_ratio=4, dropout=0.0, val=True, split=val, save_json=False, conf=None, iou=0.7, max_det=300, half=False, dnn=False, plots=True, source=None, vid_stride=1, stream_buffer=False, visualize=False, augment=False, agnostic_nms=False, classes=None, retina_masks=False, embed=None, show=False, save_frames=False, save_txt=False, save_conf=False, save_crop=False, show_labels=True, show_conf=True, show_boxes=True, line_wid

100%|██████████| 433k/433k [00:00<00:00, 14.6MB/s]
Unzipping C:\Source\datasets\coco8.zip to C:\Source\datasets\coco8...: 100%|██████████| 25/25 [00:00<00:00, 1180.40file/s]

Dataset download success  (0.9s), saved to C:\Source\datasets



100%|██████████| 755k/755k [00:00<00:00, 17.6MB/s]




                   from  n    params  module                                       arguments                     
  0                  -1  1       464  ultralytics.nn.modules.conv.Conv             [3, 16, 3, 2]                 
  1                  -1  1      4672  ultralytics.nn.modules.conv.Conv             [16, 32, 3, 2]                
  2                  -1  1      6640  ultralytics.nn.modules.block.C3k2            [32, 64, 1, False, 0.25]      
  3                  -1  1     36992  ultralytics.nn.modules.conv.Conv             [64, 64, 3, 2]                
  4                  -1  1     26080  ultralytics.nn.modules.block.C3k2            [64, 128, 1, False, 0.25]     
  5                  -1  1    147712  ultralytics.nn.modules.conv.Conv             [128, 128, 3, 2]              
  6                  -1  1     87040  ultralytics.nn.modules.block.C3k2            [128, 128, 1, True]           
  7                  -1  1    295424  ultralytics.nn.modules.conv.Conv             [12

train: Scanning C:\Source\datasets\coco8\labels\train... 4 images, 0 backgrounds, 0 corrupt: 100%|██████████| 4/4 [00:00<00:00, 3972.82it/s]

train: New cache created: C:\Source\datasets\coco8\labels\train.cache



val: Scanning C:\Source\datasets\coco8\labels\val... 4 images, 0 backgrounds, 0 corrupt: 100%|██████████| 4/4 [00:00<00:00, 3998.38it/s]

val: New cache created: C:\Source\datasets\coco8\labels\val.cache


Plotting labels to C:\Source\iot_dataanalysis_2025\runs\detect\train\labels.jpg... 
optimizer: 'optimizer=auto' found, ignoring 'lr0=0.01' and 'momentum=0.937' and determining best 'optimizer', 'lr0' and 'momentum' automatically... 
optimizer: AdamW(lr=0.000119, momentum=0.9) with parameter groups 81 weight(decay=0.0), 88 weight(decay=0.0005), 87 bias(decay=0.0)
TensorBoard: model graph visualization added 
Image sizes 640 train, 640 val
Using 8 dataloader workers
Logging results to C:\Source\iot_dataanalysis_2025\runs\detect\train
Starting training for 100 epochs...

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      1/100     0.707G      1.098       2.75      1.479         21        640: 100%|██████████| 1/1 [00:01<00:00,  1.14s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  4.30it/s]

                   all          4         17      0.567       0.85      0.878      0.634



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      2/100      0.73G      1.163      2.778      1.474         36        640: 100%|██████████| 1/1 [00:00<00:00,  3.96it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00, 24.91it/s]

                   all          4         17      0.557       0.85      0.887      0.618



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      3/100     0.736G      1.089      2.497      1.216         20        640: 100%|██████████| 1/1 [00:00<00:00, 11.08it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00, 25.41it/s]

                   all          4         17      0.559       0.85       0.85      0.615



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      4/100     0.746G      1.164      2.793      1.485         21        640: 100%|██████████| 1/1 [00:00<00:00, 10.80it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00, 23.38it/s]

                   all          4         17      0.543       0.85      0.858      0.637



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      5/100     0.746G       1.21      2.467      1.617         19        640: 100%|██████████| 1/1 [00:00<00:00,  6.33it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00, 22.35it/s]

                   all          4         17      0.548       0.85      0.858      0.637



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      6/100     0.779G     0.8327      1.918      1.148         22        640: 100%|██████████| 1/1 [00:00<00:00,  9.87it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00, 26.52it/s]

                   all          4         17      0.544      0.859      0.858      0.629



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      7/100     0.799G      1.407      2.753      1.768         20        640: 100%|██████████| 1/1 [00:00<00:00, 11.00it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00, 26.34it/s]

                   all          4         17      0.534      0.864      0.859      0.629



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      8/100     0.809G      1.241      3.857      1.692         20        640: 100%|██████████| 1/1 [00:00<00:00, 10.85it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00, 22.89it/s]

                   all          4         17      0.557      0.867      0.863      0.631



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      9/100     0.822G     0.9127      2.976      1.275         20        640: 100%|██████████| 1/1 [00:00<00:00,  9.35it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00, 25.28it/s]

                   all          4         17      0.544      0.867      0.897      0.633



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     10/100      0.84G       1.08      2.198      1.296         25        640: 100%|██████████| 1/1 [00:00<00:00,  8.88it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00, 22.92it/s]

                   all          4         17      0.533      0.867      0.895      0.642



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     11/100     0.854G      1.173      2.263      1.557         31        640: 100%|██████████| 1/1 [00:00<00:00,  8.78it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00, 22.69it/s]

                   all          4         17      0.644      0.944      0.912       0.65



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     12/100     0.869G      1.014      2.356      1.396         31        640: 100%|██████████| 1/1 [00:00<00:00, 10.49it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00, 20.46it/s]

                   all          4         17      0.649      0.944      0.911      0.645



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     13/100     0.885G      1.366      3.254      1.807         24        640: 100%|██████████| 1/1 [00:00<00:00,  9.42it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00, 21.69it/s]

                   all          4         17       0.63       0.95      0.912      0.648



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     14/100     0.898G     0.8433      2.457      1.262         15        640: 100%|██████████| 1/1 [00:00<00:00,  9.82it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00, 25.80it/s]

                   all          4         17       0.63       0.95      0.913      0.648



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     15/100     0.914G      1.271       3.04      1.525         38        640: 100%|██████████| 1/1 [00:00<00:00,  9.57it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00, 25.61it/s]

                   all          4         17      0.633      0.783       0.91      0.652



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     16/100     0.928G      1.211       2.04      1.509         49        640: 100%|██████████| 1/1 [00:00<00:00,  9.73it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00, 21.88it/s]

                   all          4         17      0.634      0.783       0.91       0.65



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     17/100     0.943G     0.8203      1.549      1.184         25        640: 100%|██████████| 1/1 [00:00<00:00, 10.34it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00, 23.96it/s]

                   all          4         17      0.641      0.783      0.911      0.629



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     18/100     0.963G     0.6249      1.342     0.9999         16        640: 100%|██████████| 1/1 [00:00<00:00, 12.53it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00, 20.81it/s]

                   all          4         17      0.641      0.783      0.911      0.629



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     19/100      0.98G     0.7282      1.613      1.054         34        640: 100%|██████████| 1/1 [00:00<00:00,  8.53it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00, 22.93it/s]

                   all          4         17      0.823       0.65      0.872       0.64



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     20/100     0.988G      1.177      1.758       1.35         25        640: 100%|██████████| 1/1 [00:00<00:00, 11.39it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00, 22.95it/s]

                   all          4         17      0.823       0.65      0.872       0.64



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     21/100      1.01G     0.7776      1.808      1.186         26        640: 100%|██████████| 1/1 [00:00<00:00, 10.30it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00, 23.51it/s]

                   all          4         17      0.789       0.65      0.856      0.627



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     22/100      1.02G      1.074        1.7      1.246         52        640: 100%|██████████| 1/1 [00:00<00:00, 12.29it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00, 25.13it/s]

                   all          4         17      0.789       0.65      0.856      0.627



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     23/100      1.04G     0.9644      1.743      1.406         22        640: 100%|██████████| 1/1 [00:00<00:00,  8.06it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00, 21.14it/s]

                   all          4         17      0.778       0.65      0.856      0.627



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     24/100      1.05G     0.9311      1.658      1.233         34        640: 100%|██████████| 1/1 [00:00<00:00, 10.91it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00, 24.66it/s]

                   all          4         17      0.778       0.65      0.856      0.627



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     25/100      1.07G     0.6523      2.014      1.068         11        640: 100%|██████████| 1/1 [00:00<00:00,  8.64it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00, 23.77it/s]

                   all          4         17      0.778       0.65      0.856      0.608



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     26/100      1.08G     0.8513      1.312      1.179         35        640: 100%|██████████| 1/1 [00:00<00:00, 13.33it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00, 24.61it/s]

                   all          4         17      0.778       0.65      0.856      0.608



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     27/100       1.1G     0.9637       1.47      1.404         24        640: 100%|██████████| 1/1 [00:00<00:00,  7.81it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00, 23.99it/s]

                   all          4         17      0.761       0.65      0.855      0.596



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     28/100      1.11G      0.965      1.211      1.248         35        640: 100%|██████████| 1/1 [00:00<00:00, 12.77it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00, 22.53it/s]

                   all          4         17      0.761       0.65      0.855      0.596



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     29/100      1.13G     0.9284      1.656      1.281         28        640: 100%|██████████| 1/1 [00:00<00:00,  8.92it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00, 23.56it/s]

                   all          4         17      0.763       0.65      0.855      0.608



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     30/100      1.14G      1.035      1.272      1.322         29        640: 100%|██████████| 1/1 [00:00<00:00, 13.00it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00, 26.04it/s]

                   all          4         17      0.763       0.65      0.855      0.608



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     31/100      1.16G      1.009      1.692      1.347         25        640: 100%|██████████| 1/1 [00:00<00:00,  8.32it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00, 24.32it/s]

                   all          4         17      0.745       0.65      0.855      0.578



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     32/100      1.17G     0.6691      1.512      1.309         20        640: 100%|██████████| 1/1 [00:00<00:00, 13.53it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00, 24.44it/s]

                   all          4         17      0.745       0.65      0.855      0.578



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     33/100      1.19G      1.045      1.949      1.348         31        640: 100%|██████████| 1/1 [00:00<00:00,  8.97it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00, 23.75it/s]

                   all          4         17      0.771       0.65      0.861      0.568



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     34/100       1.2G      1.075      1.947      1.457         32        640: 100%|██████████| 1/1 [00:00<00:00, 12.85it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00, 25.78it/s]

                   all          4         17      0.771       0.65      0.861      0.568



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     35/100      1.22G     0.9153      1.245      1.311         23        640: 100%|██████████| 1/1 [00:00<00:00,  9.33it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00, 22.65it/s]

                   all          4         17      0.783       0.65      0.852      0.564



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     36/100      1.23G      1.023       1.47      1.386         34        640: 100%|██████████| 1/1 [00:00<00:00, 12.90it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00, 23.46it/s]

                   all          4         17      0.783       0.65      0.852      0.564



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     37/100      1.25G     0.9758      1.044      1.337         34        640: 100%|██████████| 1/1 [00:00<00:00,  9.59it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00, 19.66it/s]

                   all          4         17      0.795       0.65      0.849      0.562



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     38/100      1.26G     0.5493     0.8292     0.9508         29        640: 100%|██████████| 1/1 [00:00<00:00, 11.13it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00, 24.67it/s]

                   all          4         17      0.795       0.65      0.849      0.562



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     39/100      1.28G     0.7212     0.7923      1.174         25        640: 100%|██████████| 1/1 [00:00<00:00,  9.54it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00, 22.37it/s]

                   all          4         17      0.716       0.65      0.851      0.547



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     40/100      1.29G     0.7664      1.003      1.164         19        640: 100%|██████████| 1/1 [00:00<00:00, 10.77it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00, 22.20it/s]

                   all          4         17      0.716       0.65      0.851      0.547



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     41/100      1.31G     0.7671      1.106      1.122         29        640: 100%|██████████| 1/1 [00:00<00:00,  9.34it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00, 23.33it/s]

                   all          4         17      0.699       0.65       0.85      0.549



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     42/100      1.32G      0.896       1.09      1.177         34        640: 100%|██████████| 1/1 [00:00<00:00, 13.12it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00, 26.27it/s]

                   all          4         17      0.699       0.65       0.85      0.549



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     43/100      1.34G      1.001     0.9861      1.479         20        640: 100%|██████████| 1/1 [00:00<00:00,  9.70it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00, 26.21it/s]

                   all          4         17      0.776       0.65      0.848      0.527



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     44/100      1.35G     0.6848     0.9406      1.084         23        640: 100%|██████████| 1/1 [00:00<00:00,  9.79it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00, 21.44it/s]

                   all          4         17      0.776       0.65      0.848      0.527



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     45/100      1.37G     0.6584      1.435       1.09         40        640: 100%|██████████| 1/1 [00:00<00:00,  7.96it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00, 23.07it/s]

                   all          4         17      0.798       0.65      0.847      0.527



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     46/100      1.38G     0.8697      1.017      1.258         23        640: 100%|██████████| 1/1 [00:00<00:00, 11.45it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00, 21.76it/s]

                   all          4         17      0.798       0.65      0.847      0.527



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     47/100       1.4G     0.8764     0.7681      1.263         14        640: 100%|██████████| 1/1 [00:00<00:00,  9.32it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00, 23.10it/s]

                   all          4         17      0.819      0.631      0.848      0.529



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     48/100      1.41G     0.7879      1.029      1.148         31        640: 100%|██████████| 1/1 [00:00<00:00, 12.35it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00, 21.93it/s]

                   all          4         17      0.819      0.631      0.848      0.529



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     49/100      1.43G     0.9999      1.108      1.374         17        640: 100%|██████████| 1/1 [00:00<00:00,  8.76it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00, 20.96it/s]

                   all          4         17      0.802      0.649      0.766      0.452



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     50/100      1.44G     0.6589     0.8777      1.103         31        640: 100%|██████████| 1/1 [00:00<00:00, 12.56it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00, 26.30it/s]

                   all          4         17      0.802      0.649      0.766      0.452



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     51/100      1.46G     0.7199     0.9254      1.211         35        640: 100%|██████████| 1/1 [00:00<00:00,  9.14it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00, 23.13it/s]

                   all          4         17      0.734      0.567      0.673      0.386



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     52/100      1.47G     0.8262      1.654       1.19         55        640: 100%|██████████| 1/1 [00:00<00:00, 13.92it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00, 26.56it/s]

                   all          4         17      0.734      0.567      0.673      0.386



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     53/100       1.5G     0.6383      1.049       1.08         14        640: 100%|██████████| 1/1 [00:00<00:00, 12.77it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00, 24.98it/s]

                   all          4         17      0.734      0.567      0.673      0.386



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     54/100      1.51G     0.7169     0.9855      1.103         46        640: 100%|██████████| 1/1 [00:00<00:00,  9.57it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00, 17.61it/s]

                   all          4         17      0.791      0.483      0.683      0.359



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     55/100      1.51G     0.6532      1.136      1.145         30        640: 100%|██████████| 1/1 [00:00<00:00, 12.63it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00, 21.72it/s]

                   all          4         17      0.791      0.483      0.683      0.359



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     56/100      1.54G     0.6833      0.849      1.217         22        640: 100%|██████████| 1/1 [00:00<00:00, 11.58it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00, 24.03it/s]

                   all          4         17      0.791      0.483      0.683      0.359



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     57/100      1.55G     0.7164     0.7322      1.212         28        640: 100%|██████████| 1/1 [00:00<00:00,  9.19it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00, 22.08it/s]

                   all          4         17      0.794      0.483      0.683      0.293



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     58/100      1.56G     0.7792      1.026      1.192         11        640: 100%|██████████| 1/1 [00:00<00:00, 11.76it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00, 19.97it/s]

                   all          4         17      0.794      0.483      0.683      0.293



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     59/100      1.59G     0.7671     0.7373      1.254         16        640: 100%|██████████| 1/1 [00:00<00:00, 11.88it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00, 25.11it/s]

                   all          4         17      0.794      0.483      0.683      0.293



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     60/100       1.6G     0.7434      1.025      1.287         30        640: 100%|██████████| 1/1 [00:00<00:00,  9.41it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00, 24.26it/s]

                   all          4         17      0.845       0.48      0.543      0.257



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     61/100       1.6G     0.6051      0.688      1.063         26        640: 100%|██████████| 1/1 [00:00<00:00,  9.81it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00, 16.11it/s]

                   all          4         17      0.845       0.48      0.543      0.257



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     62/100      1.63G     0.7546     0.9266      1.297         20        640: 100%|██████████| 1/1 [00:00<00:00, 11.73it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00, 20.38it/s]

                   all          4         17      0.845       0.48      0.543      0.257



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     63/100      1.64G     0.6662     0.8457      1.163         16        640: 100%|██████████| 1/1 [00:00<00:00,  9.28it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00, 23.43it/s]

                   all          4         17      0.856      0.472      0.532      0.248



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     64/100      1.65G     0.7963     0.7768      1.086         26        640: 100%|██████████| 1/1 [00:00<00:00, 13.75it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00, 25.62it/s]

                   all          4         17      0.856      0.472      0.532      0.248



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     65/100      1.67G     0.6187      0.695      1.042         37        640: 100%|██████████| 1/1 [00:00<00:00, 11.88it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00, 22.75it/s]

                   all          4         17      0.856      0.472      0.532      0.248



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     66/100      1.69G     0.9413      1.951      1.333         51        640: 100%|██████████| 1/1 [00:00<00:00,  8.19it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00, 23.38it/s]

                   all          4         17      0.849      0.479      0.523       0.25



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     67/100      1.69G     0.7584     0.9315      1.182         37        640: 100%|██████████| 1/1 [00:00<00:00, 12.41it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00, 22.89it/s]

                   all          4         17      0.849      0.479      0.523       0.25



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     68/100      1.72G     0.6322     0.7745      1.027         31        640: 100%|██████████| 1/1 [00:00<00:00, 12.57it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00, 24.96it/s]

                   all          4         17      0.849      0.479      0.523       0.25



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     69/100      1.73G     0.7433     0.7106       1.29         12        640: 100%|██████████| 1/1 [00:00<00:00,  8.98it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00, 23.50it/s]

                   all          4         17      0.845       0.48      0.524      0.245



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     70/100      1.74G      1.022      1.079      1.452          9        640: 100%|██████████| 1/1 [00:00<00:00, 11.53it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00, 24.97it/s]

                   all          4         17      0.845       0.48      0.524      0.245



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     71/100      1.76G     0.6662      0.703      1.152         23        640: 100%|██████████| 1/1 [00:00<00:00, 10.53it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00, 22.08it/s]

                   all          4         17      0.845       0.48      0.524      0.245



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     72/100      1.78G     0.6165     0.9729      1.152         21        640: 100%|██████████| 1/1 [00:00<00:00,  9.28it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00, 22.36it/s]

                   all          4         17      0.829      0.481       0.53       0.26



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     73/100      1.78G     0.5635     0.5976      1.116         15        640: 100%|██████████| 1/1 [00:00<00:00, 12.04it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00, 20.92it/s]

                   all          4         17      0.829      0.481       0.53       0.26



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     74/100      1.81G      1.061       1.89      1.486         28        640: 100%|██████████| 1/1 [00:00<00:00, 11.46it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00, 23.79it/s]

                   all          4         17      0.829      0.481       0.53       0.26



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     75/100      1.82G     0.9339     0.8899      1.234         60        640: 100%|██████████| 1/1 [00:00<00:00,  8.79it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00, 22.73it/s]

                   all          4         17      0.862      0.476      0.526      0.261



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     76/100      1.83G     0.9155     0.8915      1.335         14        640: 100%|██████████| 1/1 [00:00<00:00, 10.77it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00, 23.50it/s]

                   all          4         17      0.862      0.476      0.526      0.261



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     77/100      1.85G     0.6779     0.6292      1.043         37        640: 100%|██████████| 1/1 [00:00<00:00, 11.51it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00, 21.47it/s]

                   all          4         17      0.862      0.476      0.526      0.261



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     78/100      1.87G     0.6606      1.006      1.126         45        640: 100%|██████████| 1/1 [00:00<00:00,  8.66it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00, 21.19it/s]

                   all          4         17      0.855      0.459      0.523      0.257



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     79/100      1.88G     0.8996      1.456      1.344         16        640: 100%|██████████| 1/1 [00:00<00:00, 11.40it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00, 19.07it/s]

                   all          4         17      0.855      0.459      0.523      0.257



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     80/100       1.9G      0.621     0.6226       1.12         21        640: 100%|██████████| 1/1 [00:00<00:00, 11.49it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00, 23.12it/s]

                   all          4         17      0.855      0.459      0.523      0.257



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     81/100      1.91G     0.7918     0.7814      1.107         42        640: 100%|██████████| 1/1 [00:00<00:00,  7.81it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00, 20.59it/s]

                   all          4         17      0.856      0.456      0.525      0.253



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     82/100      1.92G     0.4436     0.4465     0.9404         32        640: 100%|██████████| 1/1 [00:00<00:00, 10.91it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00, 24.96it/s]

                   all          4         17      0.856      0.456      0.525      0.253



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     83/100      1.94G     0.6752     0.7486      1.039         31        640: 100%|██████████| 1/1 [00:00<00:00, 12.66it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00, 26.42it/s]

                   all          4         17      0.856      0.456      0.525      0.253



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     84/100      1.96G     0.8206      1.379      1.272         30        640: 100%|██████████| 1/1 [00:00<00:00,  8.77it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00, 22.85it/s]

                   all          4         17      0.851      0.457      0.527      0.251



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     85/100      1.96G     0.6159     0.7253      1.174         26        640: 100%|██████████| 1/1 [00:00<00:00, 11.52it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00, 21.90it/s]

                   all          4         17      0.851      0.457      0.527      0.251



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     86/100      1.99G     0.6397     0.8651      1.157         24        640: 100%|██████████| 1/1 [00:00<00:00, 11.16it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00, 25.53it/s]

                   all          4         17      0.851      0.457      0.527      0.251



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     87/100         2G     0.5913     0.6655      1.095         31        640: 100%|██████████| 1/1 [00:00<00:00, 10.34it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00, 21.19it/s]

                   all          4         17      0.851      0.457      0.527      0.251



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     88/100      2.02G     0.7254     0.9907      1.119         47        640: 100%|██████████| 1/1 [00:00<00:00,  9.78it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00, 24.65it/s]

                   all          4         17      0.809      0.444      0.511      0.245



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     89/100      2.02G     0.6493     0.7954      1.159         26        640: 100%|██████████| 1/1 [00:00<00:00, 12.08it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00, 25.28it/s]

                   all          4         17      0.809      0.444      0.511      0.245



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     90/100      2.04G     0.8045     0.6919      1.105         31        640: 100%|██████████| 1/1 [00:00<00:00, 12.25it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00, 25.07it/s]

                   all          4         17      0.809      0.444      0.511      0.245


Closing dataloader mosaic

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     91/100      2.06G      0.571     0.5492       1.06         13        640: 100%|██████████| 1/1 [00:04<00:00,  4.67s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00, 17.58it/s]

                   all          4         17      0.809      0.444      0.511      0.245



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     92/100      2.08G     0.4512     0.4808     0.8512         13        640: 100%|██████████| 1/1 [00:00<00:00, 10.60it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00, 25.21it/s]

                   all          4         17       0.86      0.456      0.521      0.252



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     93/100      2.08G     0.4633     0.4968     0.9587         13        640: 100%|██████████| 1/1 [00:00<00:00, 13.84it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00, 21.45it/s]

                   all          4         17       0.86      0.456      0.521      0.252



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     94/100      2.11G      0.609     0.6229      1.094         13        640: 100%|██████████| 1/1 [00:00<00:00, 10.71it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00, 25.94it/s]

                   all          4         17       0.86      0.456      0.521      0.252



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     95/100      2.12G     0.5564     0.6846      1.062         13        640: 100%|██████████| 1/1 [00:00<00:00, 12.56it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00, 25.94it/s]

                   all          4         17       0.86      0.456      0.521      0.252



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     96/100      2.14G     0.8399     0.7876      1.226         13        640: 100%|██████████| 1/1 [00:00<00:00,  8.69it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00, 21.44it/s]

                   all          4         17      0.861      0.458       0.52      0.267



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     97/100      2.14G      0.521      0.455     0.8951         13        640: 100%|██████████| 1/1 [00:00<00:00, 12.25it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00, 23.22it/s]

                   all          4         17      0.861      0.458       0.52      0.267



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     98/100      2.17G     0.5424     0.6244      1.008         13        640: 100%|██████████| 1/1 [00:00<00:00, 11.75it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00, 25.61it/s]

                   all          4         17      0.861      0.458       0.52      0.267



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     99/100      2.18G     0.5596     0.5313      1.007         13        640: 100%|██████████| 1/1 [00:00<00:00,  9.83it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00, 18.85it/s]

                   all          4         17      0.861      0.458       0.52      0.267



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


    100/100       2.2G     0.4525       0.45     0.9363         13        640: 100%|██████████| 1/1 [00:00<00:00,  7.99it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00, 24.97it/s]

                   all          4         17      0.852      0.463      0.525      0.276



100 epochs completed in 0.015 hours.
Optimizer stripped from C:\Source\iot_dataanalysis_2025\runs\detect\train\weights\last.pt, 5.5MB
Optimizer stripped from C:\Source\iot_dataanalysis_2025\runs\detect\train\weights\best.pt, 5.5MB

Validating C:\Source\iot_dataanalysis_2025\runs\detect\train\weights\best.pt...
Ultralytics 8.3.109  Python-3.11.9 torch-2.6.0+cu118 CUDA:0 (NVIDIA GeForce RTX 4060, 8188MiB)
YOLO11n summary (fused): 100 layers, 2,616,248 parameters, 0 gradients, 6.5 GFLOPs


                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00, 30.26it/s]


                   all          4         17      0.635      0.783       0.91      0.647
                person          3         10      0.616        0.7      0.654      0.325
                   dog          1          1      0.517          1      0.995      0.796
                 horse          1          2      0.578          1      0.995       0.65
              elephant          1          2       0.55          1      0.828      0.323
              umbrella          1          1      0.552          1      0.995      0.895
          potted plant          1          1          1          0      0.995      0.895
Speed: 0.2ms preprocess, 3.0ms inference, 0.0ms loss, 1.0ms postprocess per image
Results saved to C:\Source\iot_dataanalysis_2025\runs\detect\train


##### 이미지 예측

In [8]:
# yolo11n.pt  모델로 이미지 예측
result = model('./cat.jpg')


image 1/1 c:\Source\iot_dataanalysis_2025\day53\cat.jpg: 480x640 1 cat, 43.9ms
Speed: 7.2ms preprocess, 43.9ms inference, 2.3ms postprocess per image at shape (1, 3, 480, 640)


In [9]:
#matplot 모듈 로드
import matplotlib.pyplot as plt
from PIL import Image, ImageDraw, ImageFont

In [10]:
# 예측결과를 이미지+박스로 이미지 만들어 저장하기

# 모델이 예측한 객체(바운딩 박스 등)를 이미지 위에 그려준 결과를 NumPy 배열로 얻을 수 있어
img = result[0].plot()


# img는 OpenCV 스타일의 BGR 포맷이야 (Blue-Green-Red)
# PIL.Image.fromarray()는 RGB 포맷을 기대해
# 그래서 [..., ::-1]로 채널 순서를 뒤집어서 RGB로 바꾸는 것이야
img_pil = Image.fromarray(img[...,::-1])


img_pil.save('./predict_result.jpg')


##### OpenCV
- Opensource Computer Vision의 약자 . 실시간 컴퓨터비전(시각처리) 목적의 프로그래밍 라이브러리
- 인텔에서 2000년에 C, C++ 사용하기 위해서 개발
- 파이썬에서 사용할 수 있게 래핑
```shell
>pip install opencv-python
```

In [11]:
!pip install opencv-python


[notice] A new release of pip is available: 24.0 -> 25.0.1
[notice] To update, run: python.exe -m pip install --upgrade pip


In [12]:
import cv2
cv2.__version__

'4.11.0'

In [13]:
img2 = cv2.imread('./predict_result.jpg')
img2.shape     #(464, 640, 3) ->(height, width, channel)

(464, 640, 3)

In [15]:
# 윈도우 창 오픈

cv2.imshow('predict_result', img2)

# 키보드 입력을 기다리는 함수
# 0이면 → 무한히 기다림 (키보드 아무 키나 누를 때까지)
# cv2.waitKey(1000)처럼 쓰면 → 1000ms(1초) 동안 기다린 뒤 자동 종료
cv2.waitKey(0)


# mshow()로 띄운 모든 창을 닫음
# 여러 이미지 창을 띄웠을 때도 전부 한 번에 닫아줘요
# 안 쓰면 창이 그대로 남아 있을 수 있어요 (특히 Jupyter나 IDE에서는)
cv2.destroyAllWindows()

<img src = './cv2로윈도우창에predict_result띄우기.png' width=500>
<img src = './predict_result이미지크기(높이,너비).png' width=500>


In [17]:
#objects.jpg 이미지가 너무 커서 사이즈 조정 먼저
img = cv2.imread('./objects.jpg')
resized_img = cv2.resize(img, (640,400))


result = model(resized_img)
plots = result[0].plot()

cv2.imshow('predct_result2', plots)
cv2.waitKey(0)
cv2.destroyAllWindows()



0: 416x640 1 bottle, 1 cup, 1 bowl, 1 dining table, 1 tv, 1 mouse, 1 book, 45.9ms
Speed: 10.5ms preprocess, 45.9ms inference, 1.5ms postprocess per image at shape (1, 3, 416, 640)


<img src = './predict_result2를window창에띄우기.png' width=500>
<img src = './predict_result2이미지크기.png' width=500>

##### 동영상 플레이
- 라즈베리 웹캠 사용추천


In [24]:
video_path = './sample01.mp4'
output_path = './sample01_output.mp4'
count_path = './sample01_count.mp4'


In [19]:
#OpenCV 라이브러리를 사용해서 동영상을 재생하는 기본적인 코드

#cv2.VideoCapture는 비디오 캡처 객체를 만들어주고, 이 객체를 통해 프레임(이미지) 단위로 동영상을 읽을 수 있어요.
cap = cv2.VideoCapture(video_path)  

while cap.isOpened():   # 동영상 파일이 정상적으로 열렸는지 확인
    ret, frame = cap.read()   # 한 프레임씩 읽습니다. ret은 읽기에 성공했는지를 나타내는 불리언 값 (True 또는 False) frame은 읽어온 실제 **이미지 데이터(프레임)**입니다.
    if not ret : break

    cv2.imshow('Video Play', frame)

    if cv2.waitKey(1) & 0xFF == ord('q') :
        break
cap.release()  #비디오 캡처 객체를 해제합니다. (메모리 정리)
cv2.destroyAllWindows()

##### YOLO 실시간 예측

In [20]:
#시간 모듈 로드
import time


In [25]:
cap = cv2. VideoCapture(video_path)

fps = cap.get(cv2.CAP_PROP_FPS)
frame_time = 1.0/fps

# 영상의 프레임 너비와 높이를 가져옴 → 저장할 영상의 사이즈로 사용됨.
width = int(cap.get(cv2.CAP_PROP_FRAME_WIDTH))
height = int(cap.get(cv2.CAP_PROP_FRAME_HEIGHT))

#  영상 저장 객체 설정 (VideoWriter)
# fourcc: 동영상 코덱 설정 (여기선 'mp4v' → mp4 파일로 저장 가능)
# out: 탐지 결과를 저장할 비디오 파일 생성 (output_path에 저장됨)
fourcc = cv2.VideoWriter_fourcc(*'mp4v')
out = cv2.VideoWriter(output_path, fourcc, fps, (width, height))


while cap.isOpened():
    start_time = time.time()
    ret, frame = cap.read()
    if not ret : break

    # 객체 탐지
    results = model(frame)
    for result in results :
        detect_frame = result.plot()

    # 탐지 결과가 표시된 프레임을 새 영상 파일로 저장
    out.write(detect_frame)

    # detect_frame: 탐지 결과 표시된 영상
    cv2.imshow('YOLO Object Detection', detect_frame)
    # frame: 원본 영상 (비교용)
    cv2.imshow('Video Play', frame)

    # 프레임 속도 맞추기
    # 실제 처리에 걸린 시간 측정
    elapsed_time =time.time() -start_time  
    # 영상 원래 속도와 맞추기 위해 delay 계산 (ms 단위)  + 너무 빠르면 최소 1ms 대기
    delay = max(int((frame_time-elapsed_time)*1000),1)
   
    # 종료 조건
    if cv2.waitKey(delay) & 0xFF == ord('q') :
        break
cap.release()
out.release()
cv2.destroyAllWindows()



0: 384x640 1 train, 1 clock, 22.4ms
Speed: 3.2ms preprocess, 22.4ms inference, 1.5ms postprocess per image at shape (1, 3, 384, 640)

0: 384x640 1 train, 1 clock, 44.3ms
Speed: 2.2ms preprocess, 44.3ms inference, 8.8ms postprocess per image at shape (1, 3, 384, 640)

0: 384x640 1 train, 1 clock, 49.9ms
Speed: 3.1ms preprocess, 49.9ms inference, 13.4ms postprocess per image at shape (1, 3, 384, 640)

0: 384x640 1 train, 1 clock, 7.6ms
Speed: 3.6ms preprocess, 7.6ms inference, 1.0ms postprocess per image at shape (1, 3, 384, 640)

0: 384x640 1 train, 7.8ms
Speed: 1.1ms preprocess, 7.8ms inference, 1.1ms postprocess per image at shape (1, 3, 384, 640)

0: 384x640 1 train, 7.7ms
Speed: 1.1ms preprocess, 7.7ms inference, 1.4ms postprocess per image at shape (1, 3, 384, 640)

0: 384x640 1 train, 7.5ms
Speed: 1.1ms preprocess, 7.5ms inference, 1.1ms postprocess per image at shape (1, 3, 384, 640)

0: 384x640 1 train, 1 clock, 10.4ms
Speed: 2.1ms preprocess, 10.4ms inference, 1.4ms postproces

##### Car counting
- 지정된 라인 아래로 내려오는 자동차 개수 카운팅

In [23]:
import cv2
from ultralytics import solutions

In [28]:
# shapely 설치
!pip install shapely==2.0.1

   ---------------------------------------- 0.0/1.4 MB ? eta -:--:--
   -- ------------------------------------- 0.1/1.4 MB 1.3 MB/s eta 0:00:01
   ----------- ---------------------------- 0.4/1.4 MB 4.1 MB/s eta 0:00:01
   ------------------------ --------------- 0.8/1.4 MB 5.9 MB/s eta 0:00:01
   ---------------------------------------  1.4/1.4 MB 7.9 MB/s eta 0:00:01
   ---------------------------------------- 1.4/1.4 MB 7.2 MB/s eta 0:00:00



[notice] A new release of pip is available: 24.0 -> 25.0.1
[notice] To update, run: python.exe -m pip install --upgrade pip


In [32]:
#lap 설치치
!pip install lap

   ---------------------------------------- 0.0/1.5 MB ? eta -:--:--
   - -------------------------------------- 0.0/1.5 MB 991.0 kB/s eta 0:00:02
   -------- ------------------------------- 0.3/1.5 MB 3.8 MB/s eta 0:00:01
   ------------------ --------------------- 0.7/1.5 MB 5.5 MB/s eta 0:00:01
   ---------------------------------- ----- 1.3/1.5 MB 7.3 MB/s eta 0:00:01
   ---------------------------------------- 1.5/1.5 MB 7.2 MB/s eta 0:00:00



[notice] A new release of pip is available: 24.0 -> 25.0.1
[notice] To update, run: python.exe -m pip install --upgrade pip


In [34]:
cap = cv2.VideoCapture(video_path)

fps = cap.get(cv2.CAP_PROP_FPS) #동영상 FPS
width = int(cap.get(cv2.CAP_PROP_FRAME_WIDTH))
height= int(cap.get(cv2.CAP_PROP_FRAME_HEIGHT))


line_points = [(20,400),(1080,400)]

#videowriter 객체 생성( 동영상 화면에 그림, 글자를 그리기 위한 객체)
fourcc = cv2.VideoWriter_fourcc(*'mp4v')
out = cv2.VideoWriter(count_path, fourcc, fps, (width, height))

#물체 카운터 초기화
counter = solutions.ObjectCounter(
    show = True ,# 처리하는 동안 디스플레이 여부
    region = line_points,
    model = 'yolo11n.pt'    #상대 경로 (현재 디렉토리)  . 내부적으로 ./yolo11n.pt로 해석함

)

# 만약 코드 실행 중에 작업 디렉토리 (cwd)가 바뀌거나,
# 혹은 경로가 복잡한 위치에 있을 경우:
# 'yolo11n.pt'는 못 찾을 수 있음
# './yolo11n.pt' 또는 os.path.abspath()로 명확하게 써야 오류를 피할 수 있음
# import os
# model_path = os.path.abspath('yolo11n.pt')



while cap.isOpened():
    ret, frame = cap.read()
    if not ret : break      #자동으로 창이 닫히게

    #카운팅 결과를 원본에 바로 그림
    results = counter(frame) 
    out.write(results.plot_im)
    

cap.release()
out.release()
cv2.destroyAllWindows()

Ultralytics Solutions:  {'region': [(20, 400), (1080, 400)], 'show_in': True, 'show_out': True, 'colormap': None, 'up_angle': 145.0, 'down_angle': 90, 'kpts': [6, 8, 10], 'analytics_type': 'line', 'json_file': None, 'records': 5, 'show': True, 'model': 'yolo11n.pt', 'line_width': 3}

0: 384x640 1 train, 15.8ms
Speed: 3.4ms preprocess, 15.8ms inference, 1.3ms postprocess per image at shape (1, 3, 384, 640)
 Results: SolutionResults(classwise_count={'train': {'IN': 0, 'OUT': 0}}, total_tracks=1)

0: 384x640 1 train, 25.2ms
Speed: 5.7ms preprocess, 25.2ms inference, 12.5ms postprocess per image at shape (1, 3, 384, 640)
 Results: SolutionResults(classwise_count={'train': {'IN': 0, 'OUT': 0}}, total_tracks=1)

0: 384x640 1 train, 18.4ms
Speed: 2.0ms preprocess, 18.4ms inference, 1.2ms postprocess per image at shape (1, 3, 384, 640)
 Results: SolutionResults(classwise_count={'train': {'IN': 0, 'OUT': 0}}, total_tracks=1)

0: 384x640 1 train, 8.4ms
Speed: 1.2ms preprocess, 8.4ms inference, 1